### PR TITLE
K Tree in c++ 

### PR DESCRIPTION
**Pull request by yr7000**
…-plus-plus

### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

### Description
Added the K Tree c++ code with detailed explanation under dp section of c-plus-plus.

### Checklist

- [ yes] I've read the contribution guidelines.
- [ yes] I've checked the issue list before deciding what to submit.
- [ yes] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests
Thanks for assigning this issue(#2691 ) to me.

### CODE REVIEW

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+So whenever we have to deal all the possible choises and pick the best out of them
+then we can use recusrsion . but due to overlapping subproblems it wont be an efficient
+approach.To optimize recusive solution we use dynamic programming.
+
+**dcaught variable says whether there is atleast one edge with weight atleast d in the current path 
+(dp state)
+dp[i][j] means the number of paths from root with sum of edges is i and truth value
+of :there is atleast one edge with weight atleast d is j
+so basically our answer is dp[n][dcaught] (according to dp state).
+
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp]: Modularize this: Keep it short, sweet and minimal

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#include<bits/stdc++.h>
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp]: Keep the description above any code

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+   int ans = Number_of_Paths(n,k,d,false);
+   cout<<ans<<endl;
+    return 0;
+}
+
+/*
+Time complexity:O(n*k)
+Input: 4 3 2
+output:6
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp]: Add a new line at the EOF

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    // according to dp state the required answer will be at dp[n][dcaught]
+    return dp[n][dcaught] = cnt;
+}
+   
+
+
+int main(){
+   // memset is used to fill the dp matrix with -1
+   // why -1? it's not a possible answer in this problem
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp]: use multi-line comment

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    // if yes then simply return the answer
+    if(dp[n][dcaught]!=-1){
+        return dp[n][dcaught];
+    }
+    // cnt variable has the number of paths
+    int cnt=0;
+    // min(n,k) because including edges with weight greater than n is useless
+    for(int weight=1;weight<=min(n,k);weight++){
+        // if weight is >=d then we pass decaught further as true
+        // else we pass its current value only
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp]: use multi-line comment

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+int dp[105][2];
+int Number_of_Paths(int n,int k,int d,bool dcaught){
+    // base case: found a path with required sum of edges
+    // we need to return 1 if dcaught is true
+    // else it cant be included in the answer
+    if(n==0){
+        return dcaught==true;
+    }
+    // memoization step: checking whether the current required answer is already caluculated or not
+    // if yes then simply return the answer
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp]: use multi-line comment

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by yr7000** [comment]
Thanks, @HarshCasper for pointing out the mistakes. I made all the changes requested.
Thank you.

**Comment by yr7000** [comment]
@HarshCasper, if there are any more mistakes please point out them. I will definitely make the changes.
Thank you

File: C-Plus-Plus/dp/.cph/.K-Tree.cpp_bab2a2ac0c1f2d8dc74838fe1c040e14.prob [hunk 7]
```diff
@@ -0,0 +1 @@
+{"name":"Local: K-Tree","url":"d:\\gssoc21\\NeoAlgo\\C-Plus-Plus\\dp\\K-Tree.cpp","tests":[{"id":1615524509487,"input":"4 3 2","output":""}],"interactive":false,"memoryLimit":1024,"timeLimit":3000,"srcPath":"d:\\gssoc21\\NeoAlgo\\C-Plus-Plus\\dp\\K-Tree.cpp","group":"local","local":true}
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/.cph/.K-Tree.cpp_bab2a2ac0c1f2d8dc74838fe1c040e14.prob]: Remove this

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+So whenever we have to deal all the possible choises and pick the best out of them
+then we can use recusrsion . but due to overlapping subproblems it wont be an efficient
+approach.To optimize recusive solution we use dynamic programming.
+
+**dcaught variable says whether there is atleast one edge with weight atleast d in the current path 
+(dp state)
+dp[i][j] means the number of paths from root with sum of edges is i and truth value
+of :there is atleast one edge with weight atleast d is j
+so basically our answer is dp[n][dcaught] (according to dp state).
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp]: Try to use a more standard language. Check how to write technical documentation. We don't want you to be skilled as a native English speaker, but focus on conveying the problem idea in a minimal way

File: C-Plus-Plus/dp/K-Tree.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    memset(dp, -1, sizeof(dp));
+    int n, k, d;
+    cin >> n >> k >> d;
+    int ans = Number_of_Paths(n, k, d, false);
+    cout << ans << endl;
+    return 0;
+}
+
+/*
+Time complexity:O(n*k)
```
Comment (HarshCasper) [inline C-Plus-Plus/dp/K-Tree.cpp:67]: Add Space Complexity

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**